File: src/utils/assistantLogger.ts
export type LogLevel = 'debug' | 'info' | 'warn' | 'error' | 'critical'
export type AssistantAction =
  | 'chat_send'
  | 'chat_receive'
  | 'chat_error'
  | 'chat_retry'
  | 'panel_open'
  | 'panel_close'
  | 'agent_spawn'
  | 'agent_close'
  | 'agent_command'
  | 'agent_exit'
  | 'circuit_open'
  | 'circuit_close'
  | 'circuit_half_open'
  | 'health_check'
  | 'health_degraded'
  | 'health_recovery'
  | 'state_reset'
  | 'provider_call'
  | 'provider_error'
  | 'provider_timeout'
  | 'tool_execute'
  | 'tool_error'
  | 'tool_success'
  | 'session_start'
  | 'session_end'
  | 'recovery_attempt'
  | 'recovery_success'
  | 'recovery_failure'

export interface LogEntry {
  id: string
  timestamp: number
  level: LogLevel
  action: AssistantAction
  message: string
  correlationId?: string
  provider?: string
  durationMs?: number
  error?: {
    name: string
    message: string
    code?: string
    stack?: string
  }
  meta?: Record<string, unknown>
}

const LOG_LEVEL_PRIORITY: Record<LogLevel, number> = {
  debug: 0,
  info: 1,
  warn: 2,
  error: 3,
  critical: 4,
}

const MAX_LOG_ENTRIES = 500
const MIN_LEVEL: LogLevel = 'debug'

let correlationCounter = 0

function generateId(): string {
  return `log-${Date.now()}-${++correlationCounter}`
}

export function createCorrelationId(): string {
  return `corr-${Date.now()}-${++correlationCounter}`
}

function isErrorLike(
  value: unknown,
): value is { name?: string; message?: string; code?: string; stack?: string } {
  return typeof value === 'object' && value !== null && ('message' in value || 'stack' in value)
}

function serializeError(err: unknown): LogEntry['error'] {
  if (!isErrorLike(err)) {
    return { name: 'UnknownError', message: String(err) }
  }
  return {
    name: err.name || 'Error',
    message: err.message || 'Unknown error',
    code: (err as any).code ? String((err as any).code) : undefined,
    stack: err.stack,
  }
}

type LogListener = (entry: LogEntry) => void

const listeners = new Set<LogListener>()

export function onLog(listener: LogListener): () => void {
  listeners.add(listener)
  return () => listeners.delete(listener)
}

const entries: LogEntry[] = []

function addEntry(entry: LogEntry): void {
  entries.push(entry)
  if (entries.length > MAX_LOG_ENTRIES) {
    entries.splice(0, entries.length - MAX_LOG_ENTRIES)
  }
  for (const listener of listeners) {
    try {
      listener(entry)
    } catch {}
  }
  if (entry.level === 'error' || entry.level === 'critical') {
    console.error(`[Athena:${entry.action}] ${entry.message}`, entry.error || '')
  } else if (entry.level === 'warn') {
    console.warn(`[Athena:${entry.action}] ${entry.message}`)
  }
}

function log(
  level: LogLevel,
  action: AssistantAction,
  message: string,
  opts?: {
    correlationId?: string
    provider?: string
    durationMs?: number
    error?: unknown
    meta?: Record<string, unknown>
  },
): LogEntry {
  if (LOG_LEVEL_PRIORITY[level] < LOG_LEVEL_PRIORITY[MIN_LEVEL]) {
    const entry: LogEntry = { id: generateId(), timestamp: Date.now(), level, action, message }
    return entry
  }

  const entry: LogEntry = {
    id: generateId(),
    timestamp: Date.now(),
    level,
    action,
    message,
    correlationId: opts?.correlationId,
    provider: opts?.provider,
    durationMs: opts?.durationMs,
    error: opts?.error ? serializeError(opts.error) : undefined,
    meta: opts?.meta,
  }

  addEntry(entry)
  return entry
}

export const assistantLogger = {
  debug: (action: AssistantAction, message: string, opts?: Parameters<typeof log>[3]) =>
    log('debug', action, message, opts),
  info: (action: AssistantAction, message: string, opts?: Parameters<typeof log>[3]) =>
    log('info', action, message, opts),
  warn: (action: AssistantAction, message: string, opts?: Parameters<typeof log>[3]) =>
    log('warn', action, message, opts),
  error: (action: AssistantAction, message: string, opts?: Parameters<typeof log>[3]) =>
    log('error', action, message, opts),
  critical: (action: AssistantAction, message: string, opts?: Parameters<typeof log>[3]) =>
    log('critical', action, message, opts),

  getEntries: (opts?: {
    level?: LogLevel
    action?: AssistantAction
    since?: number
    limit?: number
  }): LogEntry[] => {
    let result = entries
    if (opts?.level)
      result = result.filter((e) => LOG_LEVEL_PRIORITY[e.level] >= LOG_LEVEL_PRIORITY[opts.level!])
    if (opts?.action) result = result.filter((e) => e.action === opts.action)
    if (opts?.since) result = result.filter((e) => e.timestamp >= opts.since!)
    if (opts?.limit) result = result.slice(-opts.limit)
    return result
  },

  getErrorCount: (sinceMs: number): number => {
    const cutoff = Date.now() - sinceMs
    return entries.filter(
      (e) => e.timestamp >= cutoff && (e.level === 'error' || e.level === 'critical'),
    ).length
  },

  clear: (): void => {
    entries.length = 0
  },
}

File: frontend/src/utils/assistant_logger.rs
//! Assistant logger — ported from src/utils/assistantLogger.ts
//!
//! Structured logging for Athena assistant actions with bounded storage
//! and listener support.

use std::collections::VecDeque;
use std::sync::{Arc, Mutex};

// ---------------------------------------------------------------------------
// Types
// ---------------------------------------------------------------------------

#[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord)]
pub enum LogLevel {
    Debug = 0,
    Info = 1,
    Warn = 2,
    Error = 3,
    Critical = 4,
}

impl serde::Serialize for LogLevel {
    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
    where
        S: serde::Serializer,
    {
        serializer.serialize_str(match self {
            LogLevel::Debug => "debug",
            LogLevel::Info => "info",
            LogLevel::Warn => "warn",
            LogLevel::Error => "error",
            LogLevel::Critical => "critical",
        })
    }
}

/// Named assistant action types for structured logging.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum AssistantAction {
    ChatSend,
    ChatReceive,
    ChatError,
    ChatRetry,
    PanelOpen,
    PanelClose,
    AgentSpawn,
    AgentClose,
    AgentCommand,
    AgentExit,
    CircuitOpen,
    CircuitClose,
    CircuitHalfOpen,
    HealthCheck,
    HealthDegraded,
    HealthRecovery,
    StateReset,
    ProviderCall,
    ProviderError,
    ProviderTimeout,
    ToolExecute,
    ToolError,
    ToolSuccess,
    SessionStart,
    SessionEnd,
    RecoveryAttempt,
    RecoverySuccess,
    RecoveryFailure,
}

impl AssistantAction {
    pub fn as_str(&self) -> &'static str {
        match self {
            Self::ChatSend => "chat_send",
            Self::ChatReceive => "chat_receive",
            Self::ChatError => "chat_error",
            Self::ChatRetry => "chat_retry",
            Self::PanelOpen => "panel_open",
            Self::PanelClose => "panel_close",
            Self::AgentSpawn => "agent_spawn",
            Self::AgentClose => "agent_close",
            Self::AgentCommand => "agent_command",
            Self::AgentExit => "agent_exit",
            Self::CircuitOpen => "circuit_open",
            Self::CircuitClose => "circuit_close",
            Self::CircuitHalfOpen => "circuit_half_open",
            Self::HealthCheck => "health_check",
            Self::HealthDegraded => "health_degraded",
            Self::HealthRecovery => "health_recovery",
            Self::StateReset => "state_reset",
            Self::ProviderCall => "provider_call",
            Self::ProviderError => "provider_error",
            Self::ProviderTimeout => "provider_timeout",
            Self::ToolExecute => "tool_execute",
            Self::ToolError => "tool_error",
            Self::ToolSuccess => "tool_success",
            Self::SessionStart => "session_start",
            Self::SessionEnd => "session_end",
            Self::RecoveryAttempt => "recovery_attempt",
            Self::RecoverySuccess => "recovery_success",
            Self::RecoveryFailure => "recovery_failure",
        }
    }
}

#[derive(Debug, Clone, serde::Serialize)]
pub struct LogEntry {
    pub id: String,
    pub timestamp: u64,
    pub level: LogLevel,
    pub action: String,
    pub message: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub correlation_id: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub provider: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub duration_ms: Option<u64>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub error: Option<LogError>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub meta: Option<serde_json::Value>,
}

#[derive(Debug, Clone, serde::Serialize)]
pub struct LogError {
    pub name: String,
    pub message: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub code: Option<String>,
}

pub type LogListener = Box<dyn Fn(&LogEntry) + Send + Sync>;

/// Options for filtering log entries.
#[derive(Debug, Clone, Default)]
pub struct GetEntriesOptions {
    pub level: Option<LogLevel>,
    pub action: Option<AssistantAction>,
    pub since: Option<u64>,
    pub limit: Option<usize>,
}

// ---------------------------------------------------------------------------
// Constants
// ---------------------------------------------------------------------------

const MAX_LOG_ENTRIES: usize = 500;
const MIN_LEVEL: LogLevel = LogLevel::Debug;

// ---------------------------------------------------------------------------
// Inner state
// ---------------------------------------------------------------------------

struct Inner {
    entries: VecDeque<LogEntry>,
    listeners: Vec<LogListener>,
    id_counter: u64,
}

// ---------------------------------------------------------------------------
// AssistantLogger
// ---------------------------------------------------------------------------

pub struct AssistantLogger {
    inner: Arc<Mutex<Inner>>,
}

impl Default for AssistantLogger {
    fn default() -> Self {
        Self::new()
    }
}

impl AssistantLogger {
    pub fn new() -> Self {
        Self {
            inner: Arc::new(Mutex::new(Inner {
                entries: VecDeque::with_capacity(MAX_LOG_ENTRIES),
                listeners: Vec::new(),
                id_counter: 0,
            })),
        }
    }

    fn log(&self, level: LogLevel, action: AssistantAction, message: &str, opts: LogOptions) {
        if level < MIN_LEVEL {
            return;
        }

        let mut inner = self.inner.lock().unwrap();
        inner.id_counter += 1;
        let entry = LogEntry {
            id: format!("log-{}-{}", now_ms(), inner.id_counter),
            timestamp: now_ms(),
            level,
            action: action.as_str().to_string(),
            message: message.to_string(),
            correlation_id: opts.correlation_id,
            provider: opts.provider,
            duration_ms: opts.duration_ms,
            error: opts.error,
            meta: opts.meta,
        };

        if inner.entries.len() >= MAX_LOG_ENTRIES {
            inner.entries.pop_front();
        }
        inner.entries.push_back(entry.clone());

        // Console logging at warn/error/critical
        match level {
            LogLevel::Error | LogLevel::Critical => {
                log::error!("[{}] {} — {}", entry.action, entry.level_as_str(), message)
            }
            LogLevel::Warn => {
                log::warn!("[{}] {} — {}", entry.action, entry.level_as_str(), message)
            }
            _ => {}
        }

        // Notify listeners
        for listener in &inner.listeners {
            listener(&entry);
        }
    }

    pub fn debug(&self, action: AssistantAction, message: &str, opts: LogOptions) {
        self.log(LogLevel::Debug, action, message, opts);
    }

    pub fn info(&self, action: AssistantAction, message: &str, opts: LogOptions) {
        self.log(LogLevel::Info, action, message, opts);
    }

    pub fn warn(&self, action: AssistantAction, message: &str, opts: LogOptions) {
        self.log(LogLevel::Warn, action, message, opts);
    }

    pub fn error(&self, action: AssistantAction, message: &str, opts: LogOptions) {
        self.log(LogLevel::Error, action, message, opts);
    }

    pub fn critical(&self, action: AssistantAction, message: &str, opts: LogOptions) {
        self.log(LogLevel::Critical, action, message, opts);
    }

    pub fn get_entries(&self, opts: &GetEntriesOptions) -> Vec<LogEntry> {
        let inner = self.inner.lock().unwrap();
        let mut results: Vec<LogEntry> = inner
            .entries
            .iter()
            .filter(|e| {
                if let Some(min_level) = opts.level {
                    if e.level < min_level {
                        return false;
                    }
                }
                if let Some(ref action) = opts.action {
                    if e.action != action.as_str() {
                        return false;
                    }
                }
                if let Some(since) = opts.since {
                    if e.timestamp < since {
                        return false;
                    }
                }
                true
            })
            .cloned()
            .collect();

        if let Some(limit) = opts.limit {
            let start = results.len().saturating_sub(limit);
            results = results.split_off(start);
        }

        results
    }

    pub fn get_error_count(&self, since_ms: u64) -> usize {
        let now = now_ms();
        let cutoff = now.saturating_sub(since_ms);
        let inner = self.inner.lock().unwrap();
        inner
            .entries
            .iter()
            .filter(|e| {
                e.timestamp >= cutoff
                    && (e.level == LogLevel::Error || e.level == LogLevel::Critical)
            })
            .count()
    }

    pub fn clear(&self) {
        let mut inner = self.inner.lock().unwrap();
        inner.entries.clear();
    }

    pub fn on_log(&self, listener: LogListener) {
        let mut inner = self.inner.lock().unwrap();
        inner.listeners.push(listener);
    }
}

impl LogEntry {
    fn level_as_str(&self) -> &'static str {
        match self.level {
            LogLevel::Debug => "DEBUG",
            LogLevel::Info => "INFO",
            LogLevel::Warn => "WARN",
            LogLevel::Error => "ERROR",
            LogLevel::Critical => "CRITICAL",
        }
    }
}

use crate::utils::time::now_ms;

/// Options for a single log call.
#[derive(Debug, Clone, Default)]
pub struct LogOptions {
    pub correlation_id: Option<String>,
    pub provider: Option<String>,
    pub duration_ms: Option<u64>,
    pub error: Option<LogError>,
    pub meta: Option<serde_json::Value>,
}

/// Generate a correlation ID for linking related log entries.
pub fn create_correlation_id() -> String {
    use std::sync::atomic::{AtomicU64, Ordering};
    static COUNTER: AtomicU64 = AtomicU64::new(0);
    let n = COUNTER.fetch_add(1, Ordering::Relaxed);
    format!("corr-{}-{}", now_ms(), n)
}
